pkg/controller/cloudmetrics/cloudmetrics_controller.go

```diff
@@ -77,47 +83,72 @@ var _ reconcile.Reconciler = &ReconcileCloudMetrics{}
 
 // ReconcileCloudMetrics reconciles a CloudMetrics object
 type ReconcileCloudMetrics struct {
-	client client.Client
-	scheme *runtime.Scheme
-	logger *logrus.Entry
+	client             client.Client
+	scheme             *runtime.Scheme
+	logger             *logrus.Entry
+	metricProviderList []providers.MetricsProvider
 }
 
 // Reconcile reads all redis and postgres crs periodically and reconcile metrics for these
 // resources.
 // The Controller will requeue the Request every 5 minutes constantly when RequeueAfter is set
 func (r *ReconcileCloudMetrics) Reconcile(request reconcile.Request) (reconcile.Result, error) {
-	r.logger.Info("Reconciling CloudMetrics")
+	r.logger.Info("reconciling CloudMetrics")
+	ctx := context.TODO()
 
 	// Fetch all redis crs
 	redisInstances := &integreatlyv1alpha1.RedisList{}
 	err := r.client.List(context.TODO(), redisInstances)
 	if err != nil {
 		return reconcile.Result{}, err
 	}
-	if len(redisInstances.Items) > 0 {
-		for _, redis := range redisInstances.Items {
-			r.logger.Infof("Found redis cr: %s", redis.Name)
-		}
-	} else {
-		r.logger.Info("Found no redis instances")
+	for _, redis := range redisInstances.Items {
+		r.logger.Infof("beginning to reconcile cloud metrics for redis cr: %s", redis.Name)
 	}
 
 	// Fetch all postgres crs
 	postgresInstances := &integreatlyv1alpha1.PostgresList{}
 	err = r.client.List(context.TODO(), postgresInstances)
 	if err != nil {
-		return reconcile.Result{}, err
+		r.logger.Error(err)
 	}
-	if len(postgresInstances.Items) > 0 {
-		for _, postgres := range postgresInstances.Items {
-			r.logger.Infof("Found postgres cr: %s", postgres.Name)
+	for _, postgres := range postgresInstances.Items {
+		r.logger.Infof("beginning to reconcile cloud metrics for postgres cr: %s", postgres.Name)
+		if err := r.reconcilePostgresMetrics(ctx, &postgres); err != nil {
+			r.logger.Error(err)
 		}
-	} else {
-		r.logger.Info("Found no postgres instances")
 	}
 
-	// Requeue every 5 minutes
+	// we want full control over when we scrape metrics
+	// to allow for this we only have a single requeue
+	// this ensures regardless of errors or return times
+	// all metrics are scraped and exposed at the same time
 	return reconcile.Result{
-		RequeueAfter: watchDuration * time.Second,
+		RequeueAfter: watchDuration,
 	}, nil
 }
+
+func (r *ReconcileCloudMetrics) reconcilePostgresMetrics(ctx context.Context, postgres *integreatlyv1alpha1.Postgres) error {
+	for _, p := range r.metricProviderList {
+		// only scrape metrics on supported strategies
+		if !p.SupportsStrategy(postgres.Status.Strategy) {
+			continue
+		}
+
+		// all postgres metric providers inherit the same interface
+		// scrapeMetrics returns scraped metrics output which contains a list of GenericCloudMetrics
+		scrapedMetricsOutput, err := p.ScrapeMetrics(ctx, &providers.ScrapeMetricsInput{
```

Comment:
Question/Observation:

Do we want the `MetricsProvider` interface to be as generic as it is here? We're passing down the name + namespace of the resource to make another API call to fetch the resource within the provider when we already have it. Would it not be a bit easier to deal with if we had separate interfaces like `PostgresMetricProvider`, `RedisMetricProvider` and maybe `BlobStorageMetricProvider` and then the implementations for those would be something like `AWSPostgresMetricProvider`, `AWSRedisMetricProvider` and so on for different cloud providers in the future?

I see the `Postgres`, `Redis` and `BlobStorage` CRs as the top level interfaces we deal with and then we have providers for each of those. To me it would kinda make sense that at a similar level you'd have `PostgresMetricProvider`, `RedisMetricProvider` and so on. It would mean that in future you could define what is expected at a high level in terms of metrics for a Postgres or for a Redis and you could ensure that it's the same across all cloud providers.